# #MapFailbruaryChallenge Poll Analysing
In this notebook, we select the top 28 ideas for February

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("../data/02_Poll/#MapFailbruaryChallenge.csv")

c:\python39\lib\site-packages\IPython\core\interactiveshell.py:3194: DtypeWarning: Columns (26,27,29,34,35,36,37,38,39,40,41,42,43,60,61,100) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


We get an error because the file is misformed.

## Inspection 🔍
Let's look at the header which must contain the variable names :

In [5]:
with open("../data/02_Poll/#MapFailbruaryChallenge.csv", encoding="utf8") as f:
    first_line = f.readline().strip('\n')

In [6]:
first_line

'"Horodateur","How do you ❤ each category ? (No response means \'no advice\' 😶) [""Cartography is my passion""]","How do you ❤ each category ? (No response means \'no advice\' 😶) [""You really should\'ve normalized that""]","How do you ❤ each category ? (No response means \'no advice\' 😶) [""that really should have been like 3 different maps""]","How do you ❤ each category ? (No response means \'no advice\' 😶) [#nothingMap (See [#thingsMap](https://twitter.com/search?q=%23thingsmap&src=typed_query))]","How do you ❤ each category ? (No response means \'no advice\' 😶) [...Ooooh, you wanted to map the OTHER variable..?]","How do you ❤ each category ? (No response means \'no advice\' 😶) [A hurricane track prediction map amended in sharpie (Donal Trump inspiration : https://twitter.com/marinemaps/status/1574849272121593868)]","How do you ❤ each category ? (No response means \'no advice\' 😶) [A map clearly made with ArcGIS©℗®™, QGIS©℗®™ or MapInfo©℗®™]","How do you ❤ each category ? (No resp

In [7]:
with open("../data/02_Poll/#MapFailbruaryChallenge.csv", encoding = "utf-8") as f:
    lines = f.readlines()

In [8]:
first_line = lines[0]

In [9]:
second_line = lines[1]

In [10]:
len(second_line.split(","))

101

In [11]:
len(first_line.split(","))

111

We see that data and header have different lengths. Not good ! ❌


## Transformation ⚙
First, we'll do reformatting on the header.

We'll correctly split the variable names :

In [12]:
s = first_line.split("]\",")
s

['"Horodateur","How do you ❤ each category ? (No response means \'no advice\' 😶) [""Cartography is my passion""',
 '"How do you ❤ each category ? (No response means \'no advice\' 😶) [""You really should\'ve normalized that""',
 '"How do you ❤ each category ? (No response means \'no advice\' 😶) [""that really should have been like 3 different maps""',
 '"How do you ❤ each category ? (No response means \'no advice\' 😶) [#nothingMap (See [#thingsMap](https://twitter.com/search?q=%23thingsmap&src=typed_query))',
 '"How do you ❤ each category ? (No response means \'no advice\' 😶) [...Ooooh, you wanted to map the OTHER variable..?',
 '"How do you ❤ each category ? (No response means \'no advice\' 😶) [A hurricane track prediction map amended in sharpie (Donal Trump inspiration : https://twitter.com/marinemaps/status/1574849272121593868)',
 '"How do you ❤ each category ? (No response means \'no advice\' 😶) [A map clearly made with ArcGIS©℗®™, QGIS©℗®™ or MapInfo©℗®™',
 '"How do you ❤ each cate

We reformat the variables :

In [16]:
def reformat_string(elt):
    elt = elt\
    .replace("How do you ❤ each category ? (No response means \'no advice\' 😶)", "")\
    .replace("[", "")\
    .replace("\'" , "")\
    .replace("\"" , "")\
    .strip()
    
    return(elt)

variables = [reformat_string(elt) for elt in s]
variables

['Horodateur, Cartography is my passion',
 'You really shouldve normalized that',
 'that really should have been like 3 different maps',
 '#nothingMap (See #thingsMap](https://twitter.com/search?q=%23thingsmap&src=typed_query))',
 '...Ooooh, you wanted to map the OTHER variable..?',
 'A hurricane track prediction map amended in sharpie (Donal Trump inspiration : https://twitter.com/marinemaps/status/1574849272121593868)',
 'A map clearly made with ArcGIS©℗®™, QGIS©℗®™ or MapInfo©℗®™',
 'Bad point symbols',
 'Bees',
 'Chloropleth',
 'Chuck Norris says : you do not need to know geography to be a cartographer',
 'Colorblind',
 'Contrast',
 'Could have been a bar chart',
 'Doritos Map of the US',
 'Drop shadow connoisseur',
 'Egregious and unforgivable omissions: New Zealand, Aleutian Islands, UP Michigan, Scandinavia,  etc.',
 'Historic Map With Terrain Blender Render Bender',
 'Horrible use of choropleth',
 'Im a Scientist',
 'In a hurry',
 'Irrelevant variables',
 'JeT ScaLe CAn MeLT eY

Actually, the first element contains the timestamp

In [17]:
variables[0]

'Horodateur, Cartography is my passion'

In [18]:
variables[0].split(", ")

['Horodateur', 'Cartography is my passion']

We correctly assembly the variable names :

In [19]:
variables = variables[0].split(", ") + variables[1:len(variables)]
variables

['Horodateur',
 'Cartography is my passion',
 'You really shouldve normalized that',
 'that really should have been like 3 different maps',
 '#nothingMap (See #thingsMap](https://twitter.com/search?q=%23thingsmap&src=typed_query))',
 '...Ooooh, you wanted to map the OTHER variable..?',
 'A hurricane track prediction map amended in sharpie (Donal Trump inspiration : https://twitter.com/marinemaps/status/1574849272121593868)',
 'A map clearly made with ArcGIS©℗®™, QGIS©℗®™ or MapInfo©℗®™',
 'Bad point symbols',
 'Bees',
 'Chloropleth',
 'Chuck Norris says : you do not need to know geography to be a cartographer',
 'Colorblind',
 'Contrast',
 'Could have been a bar chart',
 'Doritos Map of the US',
 'Drop shadow connoisseur',
 'Egregious and unforgivable omissions: New Zealand, Aleutian Islands, UP Michigan, Scandinavia,  etc.',
 'Historic Map With Terrain Blender Render Bender',
 'Horrible use of choropleth',
 'Im a Scientist',
 'In a hurry',
 'Irrelevant variables',
 'JeT ScaLe CAn MeLT

In [21]:
len(variables)

101

Let's remove '\n' occurrences (just in case) :

In [20]:
lines = [elt.strip('\n') for elt in lines]

Let's separate data (which starts at second line) :

In [26]:
data = lines[1:len(lines)]

In [27]:
data[0:2]

['"2022/10/17 5:49:18 PM UTC+1","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","",""',
 '"2022/10/17 5:49:19 PM UTC+1","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","","",""']

In [28]:
data = [elt.split(",") for elt in data]

In [29]:
data[0:2]

[['"2022/10/17 5:49:18 PM UTC+1"',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""'],
 ['"2022/10/17 5:49:19 PM UTC+1"',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
  '""',
 

In [31]:
len(data[0]) == len(variables)

True

It's OK : variables and data have the same length.

We can concatenate data.

In [91]:
print(len(data))

30012


By the way, we have **30012** results for the poll.

We have to transform each element of the list to make it a dictionary that will be converted afterwards as a data frame :

In [32]:
def make_dict(l):
    data2 = [[elt] for elt in l]
    d = dict(zip(variables, data2))
    return(d)

make_dict(data[0])

{'Horodateur': ['"2022/10/17 5:49:18 PM UTC+1"'],
 'Cartography is my passion': ['""'],
 'You really shouldve normalized that': ['""'],
 'that really should have been like 3 different maps': ['""'],
 '#nothingMap (See #thingsMap](https://twitter.com/search?q=%23thingsmap&src=typed_query))': ['""'],
 '...Ooooh, you wanted to map the OTHER variable..?': ['""'],
 'A hurricane track prediction map amended in sharpie (Donal Trump inspiration : https://twitter.com/marinemaps/status/1574849272121593868)': ['""'],
 'A map clearly made with ArcGIS©℗®™, QGIS©℗®™ or MapInfo©℗®™': ['""'],
 'Bad point symbols': ['""'],
 'Bees': ['""'],
 'Chloropleth': ['""'],
 'Chuck Norris says : you do not need to know geography to be a cartographer': ['""'],
 'Colorblind': ['""'],
 'Contrast': ['""'],
 'Could have been a bar chart': ['""'],
 'Doritos Map of the US': ['""'],
 'Drop shadow connoisseur': ['""'],
 'Egregious and unforgivable omissions: New Zealand, Aleutian Islands, UP Michigan, Scandinavia,  etc.':

In [ ]:
rows = []
for i, elt in enumerate(data):
    print(i)
    d = make_dict(elt)
    row = pd.DataFrame(d)
    rows.append(row)

In [93]:
data = pd.concat(rows)

Let's export our cleant file :

In [94]:
data.to_csv("../data/02_Poll/#MapFailbruaryChallenge--cleaned.csv")

In [98]:
data.columns

Index(['Horodateur', 'Cartography is my passion',
       'You really shouldve normalized that',
       'that really should have been like 3 different maps',
       '#nothingMap (See #thingsMap](https://twitter.com/search?q=%23thingsmap&src=typed_query))',
       '...Ooooh, you wanted to map the OTHER variable..?',
       'A hurricane track prediction map amended in sharpie (Donal Trump inspiration : https://twitter.com/marinemaps/status/1574849272121593868)',
       'A map clearly made with ArcGIS©℗®™, QGIS©℗®™ or MapInfo©℗®™',
       'Bad point symbols', 'Bees', 'Chloropleth',
       'Chuck Norris says : you do not need to know geography to be a cartographer',
       'Colorblind', 'Contrast', 'Could have been a bar chart',
       'Doritos Map of the US', 'Drop shadow connoisseur',
       'Egregious and unforgivable omissions: New Zealand, Aleutian Islands, UP Michigan, Scandinavia,  etc.',
       'Historic Map With Terrain Blender Render Bender',
       'Horrible use of choropleth', '

Let's look at the values of our data :

In [111]:
v = data.loc[:, data.columns != 'Horodateur'].values

In [112]:
import numpy as np
np.unique(v)

array(['""', '"I don\'t like it  😐"', '"I like it ❤"', '"I love it ❤❤❤"'],
      dtype=object)

## Scoring 📊
Let's calculate the scores :

In [126]:
data2 = data\
.replace('""', 0)\
.replace('"I don\'t like it  😐"', -1)\
.replace('"I like it ❤"', 1)\
.replace('"I love it ❤❤❤"', 3)

In [128]:
np.unique(data2.loc[:, data.columns != 'Horodateur'].values)

array([-1,  0,  1,  3], dtype=int64)

Let's calculate the score for each idea :

In [143]:
rows = []
for idea in data2.columns[1:]:
    score = data2[[idea]].sum()
    d = {'score' : score}
    row = pd.DataFrame(d)
    rows.append(row)

In [144]:
scores = pd.concat(rows)

In [45]:
scores = scores.sort_values('score', ascending = False)

## Selection ⭐
Here, we find the 28 ideas for February :

In [46]:
scores.head(28)

Unnamed: 0  score
24              Map clearly made with Microsoft Paint     37
20                                         In a hurry     33
30  My Cartography Teacher Had A Checklist And Thi...     32
40                              Source : trust me bro     32
16  Egregious and unforgivable omissions: New Zeal...     31
47  Worst sarcastic map (e.g. electricity in the R...     29
63                                       Worst colors     29
37                           Rainbow is the new black     29
7                                   Bad point symbols     29
50                                       Worst 3D Map     29
18                         Horrible use of choropleth     28
92                                        Worst title     28
33                                Not colorblind safe     27
11                                         Colorblind     27
46  Worst make a country look like its flag classi...     26
10  Chuck Norris says : you do not need to know ge...     26
54                       Worst Valentines Day Map ? 💔     26
34                           Not enough/Too much Text     26
32                                               NSFW     26
71                               Worst hand-drawn map     25
21                               Irrelevant variables     25
2   that really should have been like 3 different ...     25
13                        Could have been a bar chart     24
70  Worst font (this is easy, everybody just gets ...     23
43  The map with the highest number of hardly dete...     22
39                              Shocking correlations     22
52                               Worst Map for hiking     22
65                                  Worst decorations     22

In [149]:
scores.to_csv("../data/03_scores/scores.csv")

## 🔁 Pick the 28 ideas

In [93]:
scores = pd.read_csv("../data/03_scores/scores.csv")

Let's pick up the 28 worst ideas :

In [94]:
ideas28 = scores.sort_values('score', ascending = False).iloc[:28, ]
ideas28

Unnamed: 0  score
24              Map clearly made with Microsoft Paint     37
20                                         In a hurry     33
30  My Cartography Teacher Had A Checklist And Thi...     32
40                              Source : trust me bro     32
16  Egregious and unforgivable omissions: New Zeal...     31
47  Worst sarcastic map (e.g. electricity in the R...     29
63                                       Worst colors     29
37                           Rainbow is the new black     29
7                                   Bad point symbols     29
50                                       Worst 3D Map     29
18                         Horrible use of choropleth     28
92                                        Worst title     28
33                                Not colorblind safe     27
11                                         Colorblind     27
46  Worst make a country look like its flag classi...     26
10  Chuck Norris says : you do not need to know ge...     26
54                       Worst Valentines Day Map ? 💔     26
34                           Not enough/Too much Text     26
32                                               NSFW     26
71                               Worst hand-drawn map     25
21                               Irrelevant variables     25
2   that really should have been like 3 different ...     25
13                        Could have been a bar chart     24
70  Worst font (this is easy, everybody just gets ...     23
43  The map with the highest number of hardly dete...     22
39                              Shocking correlations     22
52                               Worst Map for hiking     22
65                                  Worst decorations     22

Let's select Valentine's day ❤ :

In [124]:
idea_valentine = ideas28[ideas28.iloc[:,0].str.contains("Valentines")].iloc[:,0].tolist()
idea_valentine

['Worst Valentines Day Map ? 💔']

Let's select the 27 other ideas and randomize them :

In [125]:
ideas_rest = ideas28[~ideas28.iloc[:,0].str.contains("Valentine")].sample(27).iloc[:,0].tolist()
ideas_rest

['Could have been a bar chart',
 'Worst 3D Map',
 'NSFW',
 'Worst decorations',
 'Worst colors',
 'Colorblind',
 'Not enough/Too much Text',
 'Bad point symbols',
 'Chuck Norris says : you do not need to know geography to be a cartographer',
 'Not colorblind safe',
 'Worst font (this is easy, everybody just gets to use papyrus 🤪)',
 'Horrible use of choropleth',
 'Shocking correlations',
 'The map with the highest number of hardly detectable errors (e.g. slight misplacements)',
 'Source : trust me bro',
 'Map clearly made with Microsoft Paint',
 'that really should have been like 3 different maps',
 'Worst Map for hiking',
 'Egregious and unforgivable omissions: New Zealand, Aleutian Islands, UP Michigan, Scandinavia,  etc.',
 'Irrelevant variables',
 'Worst sarcastic map (e.g. electricity in the Roman Empire; population per capita)',
 'Worst make a country look like its flag classification map',
 'Worst title',
 'My Cartography Teacher Had A Checklist And This Map Ticks Them All',
 'I

Let's create our list of ideas, knowing that Valentine's day ❤ is the 14th :

In [128]:
ideas28 = [ideas_rest[i] for i in range(13)]\
+ idea_valentine\
+ [ideas_rest[i] for i in range(13, len(ideas_rest))]\

ideas28

['Could have been a bar chart',
 'Worst 3D Map',
 'NSFW',
 'Worst decorations',
 'Worst colors',
 'Colorblind',
 'Not enough/Too much Text',
 'Bad point symbols',
 'Chuck Norris says : you do not need to know geography to be a cartographer',
 'Not colorblind safe',
 'Worst font (this is easy, everybody just gets to use papyrus 🤪)',
 'Horrible use of choropleth',
 'Shocking correlations',
 'Worst Valentines Day Map ? 💔',
 'The map with the highest number of hardly detectable errors (e.g. slight misplacements)',
 'Source : trust me bro',
 'Map clearly made with Microsoft Paint',
 'that really should have been like 3 different maps',
 'Worst Map for hiking',
 'Egregious and unforgivable omissions: New Zealand, Aleutian Islands, UP Michigan, Scandinavia,  etc.',
 'Irrelevant variables',
 'Worst sarcastic map (e.g. electricity in the Roman Empire; population per capita)',
 'Worst make a country look like its flag classification map',
 'Worst title',
 'My Cartography Teacher Had A Checklist 

### Export
Let's export our 28 ideas :

In [130]:
with open(r"../data/03_scores/28ideas.csv", 'w', encoding = 'utf-8') as fp:
    fp.write('\n'.join(ideas28))